### PR TITLE
SPARK-11741 Process doctests using TextTestRunner/XMLTestRunner

### PR DESCRIPTION
**Pull request by gliptak**
SPARK-11741 Process doctests using TextTestRunner/XMLTestRunner

Doctests were moved under a class to correct processing

### CODE REVIEW

**Comment by gliptak** [comment]
https://github.com/apache/spark/pull/8323
WIP
Please review. Please note that results end up as ```target/test-reports/TEST-doctest.DocTestCase*.xml```

**Comment by gliptak** [comment]
Working to improve tooling at https://github.com/xmlrunner/unittest-xml-reporting/pull/100

**Comment by gliptak** [comment]
@davies @mengxr @JoshRosen Please review. Thanks

**Comment by gliptak** [comment]
@JoshRosen Please review.

**Comment by JoshRosen** [comment]
Same comment as on your other PR: can you please write a PR description for this which explains the change and summarizes what you did?

Also, what's the status of the `xmlrunner` changes? Is there a certain minimum supported version that we should be using in our CI environment? If so, I can help to make sure that it's installed.

**Comment by gliptak** [comment]
https://github.com/xmlrunner/unittest-xml-reporting/pull/100 was committed but not yet released. The version 1.12.0 installed with https://issues.apache.org/jira/browse/SPARK-10830 works, we could update to 1.14.0 (when released).

**Comment by JoshRosen** [comment]
Will the code here need to update after a new version of that library is released or will we only have to bump in Jenkins? If we won't have to do any code changes, then I think we should go ahead and modify the other files to also use this code. Since the few lines of test runner configuration might be duplicated in a bunch of files, we could refactor them into a separate function which we call in each file.

**Comment by gliptak** [comment]
I'm only foreseeing a library version bump.

**Comment by JoshRosen** [comment]
Okay, sounds good. Do you want to go ahead and apply this change to the other PySpark files?

**Comment by gliptak** [comment]
@mengxr Locally I see a number of these doctests fail. Should failing ones be commented and JIRAs opened against them? Thanks

**Comment by gliptak** [comment]
```
This patch adds the following public classes (experimental):

    >>> class VectorAccumulatorParam(AccumulatorParam):
```
What is the recommended way to tag this class private?

**Comment by JoshRosen** [comment]
Upthread, I suggested taking the code which handles the xmlrunner imports and configuration and putting it into a reusable function which is defined in one place and called in the `__main__` method of all test suites. Could you do this instead of copy-pasting the code into every file?

**Comment by gliptak** [comment]
@JoshRosen Please review accumulators.py

File: python/pyspark/doctesthelper.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    import unittest
+import doctest
+try:
+    import xmlrunner
+except ImportError:
+    xmlrunner = None
+
+
+def run_doctests(package):
+    print(package)
```
Comment (JoshRosen) [inline python/pyspark/doctesthelper.py]: Before merging, we're going to want to remove this print statement I think.

**Comment by JoshRosen** [comment]
Those changes to `accumulators.py` are a huge improvement, so I'd go ahead and apply that same pattern across the other files.

**Comment by gliptak** [comment]
@JoshRosen A number of the doctests do fail locally too. I'm also unsure why only a single test result shows ...

**Comment by gliptak** [comment]
@JoshRosen Could you offer ointers on how to move this forward? Thanks

**Comment by gliptak** [comment]
Rebased.

**Comment by gliptak** [comment]
@JoshRosen Could you offer pointers on how to move this forward? Thanks

**Comment by gliptak** [comment]
@JoshRosen Are you still considering this for inclusion? Thanks

**Pull request closed without merging**